elastic/kibana
# Skip rule execution when it has a newer typeVersion than Kibana instance

Resolves: https://github.com/elastic/kibana/issues/166967

This PR adds a new logic to skip task execution by using the `modelVersions` of saved objects.

`schemas.create` in modelVersions are used by Kibana to validate the SO that they are attached before saving them.
So it will force developers to bump the versions whenever there is a change in the SO (add, remove etc.)

We have already saved the latestRuleVersion in ruleTypeRegistry with the following PR https://github.com/elastic/kibana/pull/171927

And with this PR, we also save the `typeVersion` in a rule on its create, update and execute actions, so we can pass the rule type version to TaskManager with `loadIndirectParams`.
Then we can compare the `latestRuleVersion` with `typeVersion` in TaskManager and skip the execution if `typeVersion` of a rule is greater than `latestRuleVersion`. This means the rule has already been updated by a newer Kibana version.

But before passing `typeVersion` to TM, we recursively call the `isCompatibleWithPreviousVersion` method of the rule versions until one of them return false (or up to version 1) then return that version to TM.

`isCompatibleWithPreviousVersion` takes the rawRule as param, so a developer can decide if the new version is compatible with the previous one on run time.

For instance: The new field `muted` is added to the rule SO. And a developer wants to if the fetched rule SO is compatible with the previous version, so they can do:

```
isCompatibleWithPreviousVersion: (rawRule) => rawRule.muted ? true : false,
```

Another change that is introduced by tis PR:
As we are replacing the indirectParams validation with version check in TM, we decided to validate only the `rule.params` rather than the whole rawRule. Therefore we return `rule.params` as `indirectParams` and `ruleType.validate.params` as `indirectParamsSchema`

Note: we do `indirectParamsSchema.validate(indirectParams)` in TM.

## To verify:
Enable skipping feature in config:
```
xpack:
 task_manager:
   requeue_invalid_tasks:
     enabled: true
     delay: 1000
     max_attempts: 10
```     

Create a rule and check if it has the `typeVersion:1`  field,

replace `typeVersion` with `typeVersion: 2` (current latest rule type version is 1) in [here](https://github.com/elastic/kibana/pull/172161/files#diff-ce9a445b40db3d05e90a5d293e8754a57308df8fd9d9cb9af9503a255d75daf4R147)

TM should be skipping the rule execution and print 2 warnings:
`Task (xxx/xxx) has a newer version(2) than expected(1)`
`Task Manager has skipped executing the Task (xxx/xxx) x times as it has invalid params.`

after 10 skips, the should be executed.

---

typeVersion field should be addedd to the rules from an older Kibana version (after upgrade) on the first execution.


![275910888-23df2344-afa1-4176-85a8-30c0e955c1cf](https://github.com/elastic/kibana/assets/92688503/eb85b93d-1e86-4947-b115-de56c87231c0)
